File: lib/util/day_clock.dart
/// 「今日」の境界・残り時間計算をまとめたヘルパー。
class DayClock {
  static DateTime startOfDay([DateTime? date]) {
    final d = date ?? DateTime.now();
    return DateTime(d.year, d.month, d.day);
  }

  static DateTime endOfDay([DateTime? date]) {
    return startOfDay(date).add(const Duration(days: 1));
  }

  static Duration remaining([DateTime? now]) {
    final n = now ?? DateTime.now();
    final r = endOfDay(n).difference(n);
    return r.isNegative ? Duration.zero : r;
  }

  static double remainingHours([DateTime? now]) =>
      remaining(now).inSeconds / 3600.0;

  /// 「6:42」形式
  static String remainingString([DateTime? now]) {
    final total = remaining(now).inSeconds;
    final h = total ~/ 3600;
    final m = (total % 3600) ~/ 60;
    return '$h:${m.toString().padLeft(2, '0')}';
  }

  static bool isSameDay(DateTime? a, DateTime b) {
    if (a == null) return false;
    return a.year == b.year && a.month == b.month && a.day == b.day;
  }

  static int daysBetween(DateTime from, DateTime to) {
    return startOfDay(to).difference(startOfDay(from)).inDays;
  }
}

File: _ios_swiftui_reference/DoLimit/Support/DayClock.swift
import Foundation

/// 「今日」の境界・残り時間計算をまとめたヘルパー。
enum DayClock {
    static var calendar: Calendar { Calendar.current }

    static func startOfDay(_ date: Date = Date()) -> Date {
        calendar.startOfDay(for: date)
    }

    static func endOfDay(_ date: Date = Date()) -> Date {
        let start = startOfDay(date)
        return calendar.date(byAdding: .day, value: 1, to: start) ?? date
    }

    /// 今日の残り秒
    static func remainingSeconds(now: Date = Date()) -> TimeInterval {
        max(0, endOfDay(now).timeIntervalSince(now))
    }

    static func remainingHours(now: Date = Date()) -> Double {
        remainingSeconds(now: now) / 3600
    }

    /// 「6:42」形式
    static func remainingString(now: Date = Date()) -> String {
        let total = Int(remainingSeconds(now: now))
        let h = total / 3600
        let m = (total % 3600) / 60
        return String(format: "%d:%02d", h, m)
    }

    static func isSameDay(_ a: Date?, _ b: Date) -> Bool {
        guard let a else { return false }
        return calendar.isDate(a, inSameDayAs: b)
    }

    /// 指定時刻（時・分）の「今日」での Date
    static func todayAt(hour: Int, minute: Int, now: Date = Date()) -> Date {
        calendar.date(bySettingHour: hour, minute: minute, second: 0, of: now) ?? now
    }
}
